### PR TITLE
ARROW-7592: [C++] Fix crashes on corrupt IPC input

### PR DESCRIPTION
**Pull request by pitrou**
Fix the following issues spotted by OSS-Fuzz:
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20115
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20117
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20124
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20126
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20127
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20133
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20135
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20139

Those are basic missing sanity checks when reading an IPC file.

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/take_internal.h [hunk 1]
```diff
@@ -583,10 +583,8 @@ class TakerImpl<IndexSequence, UnionType> : public Taker<IndexSequence> {
       indices.set_never_out_of_bounds();
 
       // Allocate temporary storage for the offsets of all valid slots
-      // NB: Overestimates required space when indices and union_array are
-      //     not null at identical positions.
       auto child_offsets_storage_size =
-          indices.length() - std::max(union_array.null_count(), indices.null_count());
+          std::accumulate(child_counts.begin(), child_counts.end(), 0);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/take_internal.h]: @bkietz This code could lead to negative buffer size. I think the new version is accurate. Can you confirm?

**Comment by pitrou** [comment]
@wesm

File: cpp/src/arrow/ipc/message.cc [hunk 2]
```diff
@@ -275,9 +275,7 @@ Status ReadMessage(int64_t offset, int32_t metadata_length, io::RandomAccessFile
   }
 
   if (flatbuffer_length == 0) {
-    // EOS
-    *message = nullptr;
-    return Status::OK();
+    return Status::Invalid("Unexpected empty message in IPC file format");
```
Comment (pitrou) [inline cpp/src/arrow/ipc/message.cc]: @wesm Is this correct? AFAIU, even if an IPC file has an EOS message at the end, it should never be read using `ReadMessage`.

File: cpp/src/arrow/compute/kernels/take_internal.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/compute/kernels/take_internal.h]: @pitrou looks good to me. Easier to understand, too. +1

File: cpp/src/arrow/ipc/message.cc (hunk 2 above)
Comment (wesm) [reply cpp/src/arrow/ipc/message.cc]: Right, I think this is fine if there is no code assuming that they get a null message back (we'd probably want to fix such code anyway based on this experience)

File: cpp/src/arrow/type.h [hunk 3]
```diff
@@ -1288,6 +1293,9 @@ class ARROW_EXPORT DictionaryType : public FixedWidthType {
 
   bool ordered() const { return ordered_; }
 
+  static Status ValidateParameters(const DataType& index_type,
+                                   const DataType& value_type);
```
Comment (wesm) [inline cpp/src/arrow/type.h]: This need not necessarily be public, but is fine otherwise

**Review by wesm — APPROVED** [review]
+1

**Comment by wesm** [comment]
Is it worthwhile to expand our fuzzer corpus based on the oss-fuzz bugs?

**Comment by pitrou** [comment]
> Is it worthwhile to expand our fuzzer corpus based on the oss-fuzz bugs?

I have no idea. oss-fuzz probably does it on its own.

> Also should we have any helper scripts to facilitate reproducing the bugs locally?

We should probably add a documentation section about that.

File: cpp/src/arrow/type.cc [hunk 4]
```diff
@@ -501,20 +501,35 @@ Status Decimal128Type::Make(int32_t precision, int32_t scale,
 // ----------------------------------------------------------------------
 // Dictionary-encoded type
 
+Status DictionaryType::ValidateParameters(const DataType& index_type,
+                                          const DataType& value_type) {
+  const bool index_type_ok = is_integer(index_type.id()) &&
+                             checked_cast<const IntegerType&>(index_type).is_signed();
+  if (!index_type_ok) {
+    return Status::TypeError("Dictionary index type should be signed integer, got ",
```
Comment (rgsl888prabhu) [inline cpp/src/arrow/type.cc:509]: Was there any reason that only signed integer is supported and not unsigned ?

Comment (pitrou) [reply cpp/src/arrow/type.cc:509]: I'm not sure about the history (@wesm could tell more), but it's specified here:
https://arrow.apache.org/docs/format/Columnar.html#dictionary-encoded-layout

> When a field is dictionary encoded, the values are represented by an array of signed integers representing the index of the value in the dictionary.

Comment (rgsl888prabhu) [reply cpp/src/arrow/type.cc:509]: Thank you for documentation.

Comment (emkornfield) [reply cpp/src/arrow/type.cc:509]: Easier cross language support I believe

Comment (wesm) [reply cpp/src/arrow/type.cc:509]: Yes that was the thinking.

Comment (kkraus14) [reply cpp/src/arrow/type.cc:509]: cc @trxcllnt who claimed that other Arrow implementations (I presume JS) didn't enforce signed types for the indices

Comment (emkornfield) [reply cpp/src/arrow/type.cc:509]: This appears to be only mentioned in [Columnar.rst](https://github.com/apache/arrow/blob/master/docs/source/format/Columnar.rst#dictionary-encoded-layout) but not in [Schema.fbs](https://github.com/apache/arrow/blob/master/format/Schema.fbs#L282).  Schema.fbs only mentioned the default is signed.

Comment (wesm) [reply cpp/src/arrow/type.cc:509]: I'm against supporting unsigned dictionary indices. The costs outweigh the benefits in my opinion. If you want to discuss further could we move this to the Arrow dev mailing list?

**Pull request closed without merging**